### PR TITLE
GECI distance

### PR DESCRIPTION
**Pull request by nepito**
- 

# Requisitos

## Checklist [Implementación en Python](https://github.com/IslasGECI/manual#implementaci%C3%B3n-en-python)

- [x] Evoca mutantes de Python con Mutmut en GitHub Actions
- [x]  Corre una prueba dummy con Pytest en GitHub Actions
- [x]  Evalúa cambio de cobertura con Codecov en GitHub Actions
- [x]  Verifica formato de Python con Black en GitHub Actions
- [x]  Agrega phony lint a Makefile para correr Flake8 y Pylint (localmente, no en GitHub Actions)

# Revisiones

## [1ra revisión](https://github.com/IslasGECI/geci_distance/pull/2#pullrequestreview-434199160) @nepito 

- [x] [Por favor ponlo en orden alfabético](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r442931129)
- [x] [¿Por qué normalizas una pdf no siempre está normilizada? ¿por qué el índice cero te da el valor máximo?](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r442937110)
- [x] [¡¡¡Loco!!! El mutante 4 cambia (x/sigma) por (x*sigma) y no marca error la prueba, eso está muy loco](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r442944458)
- [x] [el tío Bob recomienda usar assertAlmostEqual cuando evaluas reales:](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r442948273)
- [x] [Cambiando a: (...) y el `self.assertAlmostEqual(output, 0.18126924)` ya cazas dos mutantes más](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r442950624)
- [x] [Alguna vez hizo una encuesta Evaristo y ganó make mutants tal vez lo podrías cambiar para irlo probando](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r442954267)
- [x] [`.PHONY: clean install mutants tests`](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r442954522)
- [x] [Pareciera que no es necesario regresarlas, ya que las guardan en las propiedades del objeto self.norm_hist y self.bins_mid_points. Es decir ¿qué pasa si borras esta línea?](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r442957868)

## [1ra revisión](https://github.com/IslasGECI/geci_distance/pull/2#pullrequestreview-435156761) @evaristor

- [x] [Verifica el formato en GHActions](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r443735243)
- [x] [Verifica el formato en Makefile](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r443744223)
- [x] [Instala Black](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r443747146)

## [2da revisión](https://github.com/IslasGECI/geci_distance/pull/2#pullrequestreview-435205527) @evaristor

- [x] Por favor renombra el directorio `test/` por `tests/`

## [3ra Revisión](https://github.com/IslasGECI/geci_distance/pull/2#pullrequestreview-436070229) @evaristor 

- [x] [¿Podemos quitar esta línea?](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r444441904)
- [x] [Parece que aquí no usas self ¿Sería mejor como función en lugar de método?](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r444444355)
- [x] [Parece que aquí no usas self ¿Sería mejor como función en lugar de método?](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r444444562)
- [x] [Parece que aquí no usas self ¿Sería mejor como función en lugar de método?](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r444444876)
- [x] [¿Qué pasa si no entra al if? ¿Qué regresa en ese caso? ¿Es correcto que regrese self.sigma y self.beta en lugar de self?](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r444446317)
- [x] [Parece que no usas la variable error](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r444448994)
- [x] [Podríamos definir todos los atributos en el __init__?](https://github.com/IslasGECI/geci_distance/pull/2#discussion_r444452396)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM python:3
+
+RUN pip install --upgrade pip && \
+    pip install \
+    mutmut \
+    numpy \
+    pandas \
+    scipy \
+    lmfit \
```
Comment (nepito) [inline Dockerfile]: Por favor ponlo en orden alfabético

File: geci_distance/geci_distance.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        model = Model(hazard_model)
+        model.set_param_hint('sigma', value= 1, min=0)
+        model.set_param_hint('beta', value= 1, min=0)
+        params = model.make_params()
+        return model, params
+
+    def calculate_mid_points(self, bins):
+        return (bins[:-1]+bins[1:])/2
+
+    def normalize_pdf(self, pdf):
```
Comment (nepito) [inline geci_distance/geci_distance.py]: ¿Por qué normalizas una _pdf_ no siempre está normilizada? ¿por qué el índice cero te da el valor máximo?

File: geci_distance/geci_distance.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
+
+import pandas as pd
+import numpy as np
+from lmfit import Model
+from scipy import integrate
+
+def hazard_model(x,sigma,beta):
+    return (1 - np.exp(-(x/sigma)**(-beta)))
```
Comment (nepito) [inline geci_distance/geci_distance.py]: ¡¡¡Loco!!! El mutante 4 cambia `(x/sigma)` por `(x*sigma)` y no marca error la prueba, eso está muy loco

File: test/test_geci_distance.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        """
+        self.X: int = 10
+        self.Sigma = 2
+        self.Beta: int = 0
+        self.DistancesData: np.array = np.array([1,2,3,4,5,1,2,3,10,50])
+        self.N_obs_data: np.array = np.array([1,1,1,1,1,1,1,1,1,1])
+
+    def test_hazard_model(self):
+        output = hazard_model(self.X, self.Sigma, self.Beta)
+        self.assertEqual(output, 0.6321205588285577)
```
Comment (nepito) [inline test/test_geci_distance.py]: el tío Bob recomienda usar `assertAlmostEqual` cuando evaluas reales:
```suggestion
        self.assertAlmostEqual(output, 0.18126924)
```

File: test/test_geci_distance.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+
+class Test_geci_distance(unittest.TestCase):
+    def setUp(self):
+        """
+        Crea variables que se usarán en las pruebas
+        """
+        self.X: int = 10
+        self.Sigma = 2
+        self.Beta: int = 0
```
Comment (nepito) [inline test/test_geci_distance.py]: Cambiando a:
```suggestion
        self.Beta: int = 1
```
y el `self.assertAlmostEqual(output, 0.18126924)` ya cazas dos mutantes más

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,17 @@
+mutation:
```
Comment (nepito) [inline Makefile]: Alguna vez hizo una encuesta Evaristo y ganó `make mutants` tal vez lo podrías cambiar para irlo probando
```suggestion
mutants:
```

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+	mutmut run \
+		--paths-to-mutate geci_distance
+
+.PHONY: clean install mutation tests
```
Comment (nepito) [inline Makefile]:
```suggestion
.PHONY: clean install mutants tests
```

File: geci_distance/geci_distance.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            result = model.fit(self.norm_hist, params, x=self.bins_mid_points, method='nelder')
+            self.sigma = result.params['sigma'].value
+            self.beta = result.params['beta'].value
+            return self.sigma, self.beta
+
+    def calculate_histogram(self):
+        hist, bins = np.histogram(self.distances, np.linspace(0, self.width, self.n_bins))
+        self.norm_hist = self.normalize_pdf(hist)
+        self.bins_mid_points = self.calculate_mid_points(bins)
+        return self.norm_hist, self.bins_mid_points
```
Comment (nepito) [inline geci_distance/geci_distance.py:66]: Pareciera que no es necesario regresarlas, ya que las guardan en las propiedades del objeto `self.norm_hist` y  `self.bins_mid_points`. Es decir ¿qué pasa si borras esta línea?
```suggestion
```

**Review by nepito — COMMENTED** [review]
@islasfernando no puedo solicitarte cambios

File: .github/workflows/actions.yml [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+jobs:
+  actions:
+    runs-on: ubuntu-18.04
+    steps:
+    - name: Copia repositorio
+      uses: actions/checkout@v2
+    - name: Construye imagen
+      run: docker build --tag islasgeci/geci_distance .
+    - name: Ejecuta objetivo principal del Makefile
+      run: docker run islasgeci/geci_distance make
```
Comment (devarops) [inline .github/workflows/actions.yml]:
```suggestion
      run: docker run islasgeci/geci_distance make
    - name: Verifica el formato
      run: docker run islasgeci/dummy_transformations make format
```

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+install:
+	pip install --editable .
+
```
Comment (devarops) [inline Makefile:22]:
```suggestion

format:
	black --check geci_distance
	black --check tests

```

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,12 @@
+FROM python:3
+
+RUN pip install --upgrade pip && \
+    pip install \
+    mutmut \
```
Comment (devarops) [inline Dockerfile:9]:
```suggestion
    black \
    mutmut \
```

**Review by devarops — CHANGES_REQUESTED** [review]
Por favor agrega verificación automática del formato mediante Black

**Review by devarops — CHANGES_REQUESTED** [review]
Por favor renombra el directorio `test/` por `tests/`

File: geci_distance/geci_distance.py (hunk 2 above)
Comment (islasfernando) [reply geci_distance/geci_distance.py]: creo que el nombre puede ser el error, en realidad NO es una PDF estrictamente hablando. Puede quedar mejor `normalize_histogram`. El indice 0 es el valor máximo de acuerdo a la suposición de que la probabilidad de detección cerca de la línea es 100%.

**Comment by devarops** [comment]
¿Sabes que significa esto. @islasfernando?

image

**Comment by islasfernando** [comment]
> ¿Sabes que significa esto. @islasfernando?
> 
> image

No sé, porque salen?

**Comment by devarops** [comment]
> No sé, porque salen?

Le pasé un par de _linters_ (flake8 y pylint), pero no estoy seguro sobre como interpretar los resultados

**Comment by nepito** [comment]
> ¿Sabes que significa esto. @islasfernando?
> 
> image

Nooooooo ¿te dice que andas de fantoche cargando cosas que no ocupas?

**Comment by devarops** [comment]
> Nooooooo ¿te dice que andas de fantoche cargando cosas que no ocupas?

@islasfernando: ¿No usas `pandas` en `geci_distance.py`?

**Comment by islasfernando** [comment]
> > Nooooooo ¿te dice que andas de fantoche cargando cosas que no ocupas?
> 
> @islasfernando: ¿No usas `pandas` en `geci_distance.py`?

No, ya chequé

**Comment by devarops** [comment]
> > ¿No usas `pandas` en `geci_distance.py`?
> 
> No, ya chequé

Entonces tal vez sí sean de utilidad los _linters_?

File: geci_distance/geci_distance.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
+
+import pandas as pd
```
Comment (devarops) [inline geci_distance/geci_distance.py]:
```suggestion
```

Comment (devarops) [reply geci_distance/geci_distance.py]: ¿Podemos quitar esta línea?

File: geci_distance/geci_distance.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    def set_line_length(self, length):
+        self.length = length
+
+    def set_study_area(self, area, units="km2"):
+        if units == "km2":
+            self.area = area * 1_000_000
+        if units == "m2":
+            self.area = area
+
+    def initialize_hazard_model(self):
```
Comment (devarops) [inline geci_distance/geci_distance.py]: Parece que aquí no usas `self` ¿Sería mejor como función en lugar de método?

File: geci_distance/geci_distance.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        model = Model(hazard_model)
+        model.set_param_hint("sigma", value=1, min=0)
+        model.set_param_hint("beta", value=1, min=0)
+        params = model.make_params()
+        return model, params
+
+    def calculate_mid_points(self, bins):
```
Comment (devarops) [inline geci_distance/geci_distance.py]: Parece que aquí no usas `self` ¿Sería mejor como función en lugar de método?

File: geci_distance/geci_distance.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        return (bins[:-1] + bins[1:]) / 2
+
+    def normalize_histogram(self, pdf):
```
Comment (devarops) [inline geci_distance/geci_distance.py]: Parece que aquí no usas `self` ¿Sería mejor como función en lugar de método?

File: geci_distance/geci_distance.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        return pdf / pdf[0]
+
+    def fit_detection_function(self, model="hazard"):
+        if model == "hazard":
+            model, params = self.initialize_hazard_model()
+            result = model.fit(self.norm_hist, params, x=self.bins_mid_points, method="nelder")
+            self.sigma = result.params["sigma"].value
+            self.beta = result.params["beta"].value
+            return self.sigma, self.beta
```
Comment (devarops) [inline geci_distance/geci_distance.py]: ¿Qué pasa si no entra al `if`? ¿Qué regresa en ese caso? ¿Es correcto que regrese `self.sigma` y `self.beta` en lugar de `self`?

File: geci_distance/geci_distance.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    def calculate_histogram(self):
+        hist, bins = np.histogram(self.distances, np.linspace(0, self.width, self.n_bins))
+        self.norm_hist = self.normalize_histogram(hist)
+        self.bins_mid_points = self.calculate_mid_points(bins)
+        return self.norm_hist, self.bins_mid_points
+
+    def calculate_detection_probability(self):
+        self.calculate_histogram()
+        self.fit_detection_function()
+        area, error = integrate.quad(hazard_model, 0, self.width, args=(self.sigma, self.beta))
```
Comment (devarops) [inline geci_distance/geci_distance.py]:
```suggestion
        area, _ = integrate.quad(hazard_model, 0, self.width, args=(self.sigma, self.beta))
```
Parece que no usas la variable `error`

File: geci_distance/geci_distance.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+def hazard_model(x, sigma, beta):
+    return 1 - np.exp(-((x / sigma) ** (-beta)))
+
+
+class GECI_Distance:
+    def __init__(self, n_obs, distances, n_bins=10):
+        self.distances = distances
+        self.n_bins = n_bins
+        self.n_obs = n_obs
```
Comment (devarops) [inline geci_distance/geci_distance.py]:
```suggestion
        self.area = None
        self.beta = None
        self.bins_mid_points = None
        self.detection_probability = None
        self.distances = distances
        self.length = None
        self.n_bins = n_bins
        self.n_obs = n_obs
        self.n_total = None
        self.norm_hist = None
        self.sigma = None
        self.width = None
```

Podríamos definir todos los atributos en el `__init__`?

**Review by devarops — CHANGES_REQUESTED** [review]
Excelente trabajo @islasfernando .

Aquí te dejo mi revisión

**Comment by devarops** [comment]
@islasfernando: Me llamó la atención que usemos `unittest` en lugar de `pytest`. Entiendo que `pytest` es mejor y más fácil, no? ¿Tú que opinas?

File: geci_distance/geci_distance.py (hunk 16 above)
Comment (islasfernando) [reply geci_distance/geci_distance.py]: La idea es ir agregando más modelos, y le doy la posibilidad al usuario de que guarde sigma y beta, aunque no sé si es la mejor forma de hacerlo..

Comment (devarops) [reply geci_distance/geci_distance.py]: @islasfernando 
OK. Ya entiendo. Entonces debemos regresar una excepción si el modelo seleccionado no está implementado. Buscaré como regresar excepciones en Python

Comment (devarops) [inline geci_distance/geci_distance.py]: ¿Algo así, @islasfernando ?:
```suggestion
            return self.sigma, self.beta
        else:
            raise NameError('InvalidModel')
```

Comment (devarops) [reply geci_distance/geci_distance.py]: @islasfernando : Otra opción sería quitar el `if` y nombrar esta función `def fit_detection_function_hazard(self)` y cuando en el futuro implementemos el segundo modelo (distinto a `hazard`) entonces agregamos `def fit_detection_function(self, model)` con el `if `. Para no adelantarnos al futuro.

Comment (islasfernando) [reply geci_distance/geci_distance.py]: Va, lo voy a quitar entonces. Gracias @evaristor

**Comment by nepito** [comment]
Listo @islasfernando por favor apruébala, pues yo no lo puedo hacer, pues yo soy el creador del _pull request_. Me avisas y yo hago _merge_.

**Review by islasfernando — APPROVED** [review]
